pkg/controller/operators/catalog/operator.go

```diff
@@ -1415,103 +1498,215 @@ func (o *Operator) ExecutePlan(plan *v1alpha1.InstallPlan) error {
 		case v1alpha1.StepStatusWaitingForAPI:
 			switch step.Resource.Kind {
 			case crdKind:
-				crd, err := o.opClient.ApiextensionsV1beta1Interface().ApiextensionsV1beta1().CustomResourceDefinitions().Get(step.Resource.Name, metav1.GetOptions{})
+				// First determine whether CRD is v1beta1 or v1 version
+				crdVersion, err := crdlib.Version(&step.Resource.Manifest)
 				if err != nil {
-					if k8serrors.IsNotFound(err) {
-						plan.Status.Plan[i].Status = v1alpha1.StepStatusNotPresent
-					} else {
-						return errorwrap.Wrapf(err, "error finding the %s CRD", crd.Name)
-					}
-					continue
+					return errorwrap.Wrapf(err, "error parsing step manifest: %s", step.Resource.Name)
 				}
 
-				established, namesAccepted := false, false
-				for _, cdt := range crd.Status.Conditions {
-					switch cdt.Type {
-					case v1beta1.Established:
-						if cdt.Status == v1beta1.ConditionTrue {
-							established = true
+				// Process CRD based on the version on the manifest.
+				switch crdVersion {
```

Comment:
This all looks correct, but I think it's starting to get unwieldy with 3-deep nested switches (and I think ExecutePlan is already unwieldy). 

I won't suggest you need to refactor all of ExecutePlan, but I think we can at least refactor the CRD handling.

I would suggest some interface that can manage cluster interactions based on the step. 

This is an example of what I was thinking, but any better abstraction is very welcome:

```go
type Step interface{
   Status() (error, Status)
}
type StepStatus func(error, Status)

func StepBuilder(kind, manifest, kubeClient) Step {
   // hand out concrete implementations
   if crdv1beta1: { return NewCRDV1Beta1Step(manifest, kubeClient.ApiExtensions()) }
}

func NewCRDV1Beta1Step(manifest, kubeClient.ApiExtensions()) Step {
   return func() {
       case v1alpha1.StepStatusWaitingForAPI:
           // check if apis are available
       case notexists:
          // create it
   } 
}

func ExecutePlan() error {
     for i, step := range steps {
       s := StepBuilder(step.Resource.Kind, step.Resource.Manifest)
       err, status := s.Status()
       // handle error
       steps[i].Status = status
     } 
}
```

hopefully my psuedogo is legible!